pallets/dapp-staking-v3/src/types.rs

```diff
@@ -588,178 +486,378 @@ where
-#[derive(Encode, Decode, MaxEncodedLen, Copy, Clone, Debug, PartialEq, Eq, TypeInfo, Default)]
-pub struct RewardInfo {
-    /// Rewards pool for accounts which have locked funds in dApp staking
-    #[codec(compact)]
-    pub participants: Balance,
-    /// Reward pool for dApps
-    #[codec(compact)]
-    pub dapps: Balance,
+/// Helper internal struct for iterating over `(era, stake amount)` pairs.
+///
+/// Due to how `AccountLedger` is implemented, few scenarios are possible when claming rewards:
+///
+/// 1. `staked` has some amount, `staked_future` is `None`
+///   * `maybe_first` is `None`, span describes the entire range
+/// 2. `staked` has nothing, `staked_future` is some and has some amount
+///   * `maybe_first` is `None`, span describes the entire range
+/// 3. `staked` has some amount, `staked_future` has some amount
+///   * `maybe_first` is `Some` and covers the `staked` entry, span describes the entire range except the first pair.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct EraStakePairIter {
+    /// Denotes whether the first entry is different than the others.
+    maybe_first: Option<(EraNumber, Balance)>,
+    /// Starting era of the span.
+    start_era: EraNumber,
+    /// Ending era of the span.
+    end_era: EraNumber,
+    /// Staked amount in the span.
+    amount: Balance,
+}
+
+impl EraStakePairIter {
+    /// Create new iterator struct for `(era, staked amount)` pairs.
+    pub fn new(
+        span: (EraNumber, EraNumber, Balance),
+        maybe_first: Option<(EraNumber, Balance)>,
+    ) -> Self {
+        if let Some((era, _amount)) = maybe_first {
+            debug_assert!(
+                span.0 == era + 1,
+                "The 'other', if it exists, must cover era preceding the span."
+            );
+        }
+
+        Self {
+            maybe_first,
+            start_era: span.0,
+            end_era: span.1,
+            amount: span.2,
+        }
+    }
+}
+
+impl Iterator for EraStakePairIter {
+    type Item = (EraNumber, Balance);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Fist cover the scenario where we have a unique first value
+        if let Some((era, amount)) = self.maybe_first.take() {
+            return Some((era, amount));
+        }
+
+        // Afterwards, just keep returning the same amount for different eras
+        if self.start_era <= self.end_era {
+            let value = (self.start_era, self.amount);
+            self.start_era.saturating_inc();
+            return Some(value);
+        } else {
+            None
+        }
+    }
 }
 
-// TODO: it would be nice to implement add/subtract logic on this struct and use it everywhere
-// we need to keep track of staking amount for periods. Right now I have logic duplication which is not good.
+/// Describes stake amount in an particular era/period.
 #[derive(Encode, Decode, MaxEncodedLen, Copy, Clone, Debug, PartialEq, Eq, TypeInfo, Default)]
 pub struct StakeAmount {
```

Comment:
I'll suggest to rename this struct ? This code seems weird for me:
```
                let mut stake_amount = self.staked;
                stake_amount.era = era + 1;
                stake_amount.period = current_period_info.number;
                stake_amount.add(amount, current_period_info.subperiod);
```
I would suggest `StakeInfo` or `StakingInfo` or even `StakingSnapshot` as you are using:
`let snapshot = self.staked;`

Reply:
Personally, I prefer `StakeAmount` since it refers to that, first and foremost.
And it's used in various different contexts so I wanted a generic name.

`StakeInfo` or `StakingInfo` sound _too similar_ to some existing storage items.